### PR TITLE
Add GPR functions and unit tests

### PR DESCRIPTION
**Pull request by vtommasini**
Partial merge of the Gaussian Process Regression (GPR) Library for predicting the radial expansion rate and radial velocity of BBH simulations, currently consisting only of the self contained ML class and functions (no eccentricity control related functions).

Contains the following:
1. Model class, handles input data setup with initialization, normalization, and denormalization functions
2. GPR model training function, which stores normalization statistics
3.  Prediction function, which returns the mean and uncertainty
4. GPR pipeline, which encompasses training, prediction, plots, and output metrics

Unit tests covering training, prediction, normalization/denormalization, and pipeline functionality using synthetic sine wave data are included for sanity check.

### CODE REVIEW

File: GPR_library.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
+# Distributed under the MIT License.
```
Comment (nilsvu) [inline GPR_library.py:1]: Please put this file in `src/SimulationSupport` and name it `gpr.py`.

File: GPR_library.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+#   7. train_gpr_model
+#   8. predict_with_gpr_model
+#   9. run_gpr_pipeline
+#  10. train_model_and_eigenvalue_analysis
+#  11. loo_predictions
+#  12. parse_test_runs
+#  13. apply_gpr_corrections
+#  14. save_gpr_corrected
+#  15. loo_crossval
+#  16. plot_loo_residuals
```
[the remark this answers is not in the document]
Comment (nilsvu) [reply GPR_library.py]: I suggest to just remove the list

File: GPR_library.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+# Gaussian Process Regression Machine Learning Function Library
```
Comment (nilsvu) [inline GPR_library.py]: It's good to have this comment at the start of the file, but make it a multiline string as is the convention in Python. Then it will show up in the documentation as well.

File: GPR_library.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+#  11. loo_predictions
+#  12. parse_test_runs
+#  13. apply_gpr_corrections
+#  14. save_gpr_corrected
+#  15. loo_crossval
+#  16. plot_loo_residuals
+
+import argparse
+
+import gpytorch
```
Comment (nilsvu) [inline GPR_library.py:10]: Add the dependencies to pyproject.toml (pytorch and gpytorch)

File: GPR_library.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+
+# Imports
```
Comment (nilsvu) [inline GPR_library.py]: Remove this comment. Then do the two things listed in the README.md:
```
black .
isort .
```

This will format the code and sort imports.

File: GPR_library.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+import torch
+
+
+### ML functions
```
Comment (nilsvu) [inline GPR_library.py]: Remove these comments. Make them docstrings instead (as you already have for this function)

File: GPR_library.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+    def __init__(self, train_x, train_y, likelihood):
+        super(GPRegressionModel, self).__init__(train_x, train_y, likelihood)
+
+        # Supports all dimensions (ie GPR can be run from 1-8 dimensions)
+        input_dim = train_x.shape[1] if train_x.dim() > 1 else 1
+
+        # Define base kernels - use a mixture of the RBF and Matern kernels
+        self.rbf_kernel = gpytorch.kernels.RBFKernel(ard_num_dims=input_dim)
+        self.matern_kernel = gpytorch.kernels.MaternKernel(
+            nu=2.5, ard_num_dims=input_dim
```
Comment (nilsvu) [inline GPR_library.py:48]: Explain this number 2.5 in a comment

File: GPR_library.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        raw_Y (numpy.ndarray): Raw output data
+
+    Returns:
+        GPRegressionModel: Trained model with the normalization parameters stored
+        gpytorch.likelihoods.GaussianLikelihood: Likelihood for the model
+    """
+    # Compute normalization parameters
+    input_mean = raw_X.mean(
+        axis=0
+    )  # (D, ) needed for multidimensions to avoid computing a single scalar mean for all columns combined
```
Comment (nilsvu) [inline GPR_library.py]: What do you mean with this comment? I don't see `(D, )` in the code

File: TestGPRLibrary.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
+# Distributed under the MIT License.
```
Comment (nilsvu) [inline TestGPRLibrary.py]: Move this file to `tests/` and name it `test_gpr.py`

**Review by nilsvu — CHANGES_REQUESTED** [review]
Very nice code and useful tests, the comments are very helpful 👍 please fix these minor comments and also address the copilot review above.

File: GPR_library.py (hunk 4 above)
Comment (nilsvu) [reply GPR_library.py:10]: This is still missing

File: GPR_library.py (hunk 7 above)
Comment (nilsvu) [reply GPR_library.py:48]: This is still missing

**Review by nilsvu — COMMENTED** [review]
Looks good. Please fix the few unresolved comments above. In particular, please explain the magic numbers. You can add comments explaining how you chose them, and give an idea how much you have tested different numbers. The idea is to give people editing the code an idea if these are finely tuned numbers that should not be touched, or barely explored at all and open for experimentation.

Then please squash all of your commits into one.

File: src/SimulationSupport/gpr.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+
+import gpytorch
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+
+"""
+ML functions:
+    ExactGP uses an infinite number of basis functions; GP is non-parametric and models functions globally;
+    limited only by training points
```
Comment (nilsvu) [inline src/SimulationSupport/gpr.py]: Please move this comment into the docstring for `GPRegressionModel` below. For example, write "This class derives from `gpytorch.models.ExactGP`.", then add this information you have here to explain the choice.

File: src/SimulationSupport/gpr.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+
+    # Loss function
+    mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
+
+    # Track the best model to prevent overfitting
+    best_loss = float("inf")  # Initialize best loss as infinity
+    best_state = None  # Placeholder for best model state
+
+    # Training loop for 200 iterations
+    for i in range(200):
```
Comment (nilsvu) [inline src/SimulationSupport/gpr.py:189]: Is this number 200 a finely tuned choice? Please add a comment here as well, or make the number of training iterations an input parameter (and explain how to choose a good number for it).

File: pyproject.toml [hunk 12]
```diff
@@ -13,6 +13,8 @@ requires-python = ">=3.9"
 dependencies = [
     "numpy",
     "scipy",
+    "torch",
+    "gpytorch"
```
Comment (nilsvu) [inline pyproject.toml:18]: matplotlib is missing here as well

File: tests/test_gpr.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    Assert that the shapes are reasonable, uncertainties are positive, and
+    that the mean predictions are within a reasonable range of the true values
+"""
+
+import shutil
+import unittest
+from pathlib import Path
+
+import numpy as np
+from gpr import (
```
Comment (nilsvu) [inline tests/test_gpr.py]: This line will fail when GitHub runs the test. It should be `from SimulationSupport.gpr import ...`.

**Review by nilsvu — CHANGES_REQUESTED** [review]
Please squash your commit into one (including these changes). Please also rebase on the latest `main` branch, as the test you added wasn't run by GitHub. It will run automatically once you rebased.

File: src/SimulationSupport/gpr.py [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+    # Loss function
+    mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
+
+    # Track the best model to prevent overfitting
+    best_loss = float("inf")  # Initialize best loss as infinity
+    best_state = None  # Placeholder for best model state
+
+    # Training loop for 200 iterations; 200 was chosen empirically for this dataset.
+    # In practice, early stopping terminates training once the loss
+    # plateaus, so the exact value isn't critical. If utilizing a larger
```
Comment (nilsvu) [inline src/SimulationSupport/gpr.py:185]: Where does early stopping happen? I don't see the loop below interrupted anywhere. Am I missing something?

**Review by nilsvu — APPROVED** [review]
The automated tests run now and pass 👍 Merging this now so we can move on with the next PRs. We can clarify or fix the question below in a follow-up (probably best to make the number of training iterations an input parameter with a default).

File: src/SimulationSupport/gpr.py (hunk 14 above)
Comment (vtommasini) [reply src/SimulationSupport/gpr.py:185]: fixed it now
